homeassistant/components/jewish_calendar/binary_sensor.py

```diff
@@ -53,16 +57,46 @@ def name(self):
     @property
     def is_on(self):
         """Return true if sensor is on."""
-        return self._state
+        return self._get_zmanim().issur_melacha_in_effect
 
-    async def async_update(self):
-        """Update the state of the sensor."""
-        zmanim = hdate.Zmanim(
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    def _get_zmanim(self):
+        """Return the Zmanim object for now()."""
+        return hdate.Zmanim(
             date=dt_util.now(),
             location=self._location,
             candle_lighting_offset=self._candle_lighting_offset,
             havdalah_offset=self._havdalah_offset,
             hebrew=self._hebrew,
         )
 
-        self._state = zmanim.issur_melacha_in_effect
+    async def async_added_to_hass(self):
+        """Run when entity about to be added to hass."""
+        await super().async_added_to_hass()
+        self._schedule_update()
+
+    @callback
+    def _update(self, now=None):
+        """Update the state of the sensor."""
+        self._schedule_update()
+        self.async_write_ha_state()
+
+    def _schedule_update(self):
+        """Schedule the next update of the sensor."""
+        now = dt_util.now()
+        zmanim = self._get_zmanim()
+        update = zmanim.zmanim["sunrise"] + dt.timedelta(days=1)
+        candle_lighting = zmanim.candle_lighting
+        if candle_lighting is not None and now < candle_lighting < update:
+            update = candle_lighting
+        havdalah = zmanim.havdalah
+        if havdalah is not None and now < havdalah < update:
+            update = havdalah
+        if self._next_update == update:
+            return
```

Comment:
If this case happens we'll stop updating the sensor as no new updates will be scheduled. Is that really what we want?

Reply:
The update variable is always in the future (lines 92, 94, and 97).
This condition (line 99) is a safety check to prevent re-scheduling.
However, when _schedule_update ends it's guaranteed that there is a future callback.

Reply:
We should use another guard for this. Save the return value of the event async_track_point_in_time call in an instance attribute. It can be used to remove the listener. If that attribute is truthy we know we have a scheduled listener and we should not schedule another. If we do want to schedule another we should remove the current listener first.

Reply:
Done
(The optimization was removed, and an identical time will be rescheduled.)